push-sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/apns/PushyApnsSender.java

```diff
@@ -197,93 +195,79 @@ private String createPushPayload(final Message message, final String pushMessage
+            assambleApnsClientBuilder(iOSVariant, builder);
+            connectToDestinations(iOSVariant, builder);
 
-                // APNS client has auto-reconnect, but let's log when that happens
-                apnsClient.getReconnectionFuture().addListener(future -> logger.trace("Reconnecting to APNs"));
+            // connect and wait, ONLY when we have a valid client
+            logger.debug("establishing the connection for {}", iOSVariant.getVariantID());
+            ApnsClient apnsClient;
+            try {
+                logger.debug("connecting to APNs");
+                apnsClient = builder.build();
                 return apnsClient;
+            } catch (SSLException e) {
+                logger.error("Error connecting to APNs", e);
             }
             return null;
+
         });
     }
 
-    private ApnsClient buildApnsClient(final iOSVariant iOSVariant) {
+    private void assambleApnsClientBuilder(final iOSVariant iOSVariant, final ApnsClientBuilder builder) {
 
         // this check should not be needed, but you never know:
         if (iOSVariant.getCertificate() != null && iOSVariant.getPassphrase() != null) {
 
-        // add the certificate:
-        try(final ByteArrayInputStream stream = new ByteArrayInputStream(iOSVariant.getCertificate())) {
-                final ApnsClientBuilder builder = new ApnsClientBuilder();
+            // add the certificate:
+            try (final ByteArrayInputStream stream = new ByteArrayInputStream(iOSVariant.getCertificate())) {
+
                 builder.setClientCredentials(stream, iOSVariant.getPassphrase());
 
                 if (ProxyConfiguration.hasHttpProxyConfig()) {
                     if (ProxyConfiguration.hasBasicAuth()) {
                         String user = ProxyConfiguration.getProxyUser();
                         String pass = ProxyConfiguration.getProxyPass();
-                        builder.setProxyHandlerFactory(new HttpProxyHandlerFactory(ProxyConfiguration.proxyAddress(), user, pass));
+                        builder.setProxyHandlerFactory(
+                                new HttpProxyHandlerFactory(ProxyConfiguration.proxyAddress(), user, pass));
                     } else {
                         builder.setProxyHandlerFactory(new HttpProxyHandlerFactory(ProxyConfiguration.proxyAddress()));
                     }
 
                 } else if (ProxyConfiguration.hasSocksProxyConfig()) {
                     builder.setProxyHandlerFactory(new Socks5ProxyHandlerFactory(ProxyConfiguration.socks()));
                 }
-
-                final ApnsClient apnsClient = builder.build();
-                return apnsClient;
             } catch (Exception e) {
                 logger.error("Error reading certificate", e);
-                // will be thrown below
+                // indicating an incomplete service
+                throw new IllegalArgumentException("Not able to construct APNS client");
             }
         }
```

Comment:
now the `throw` statement is moved inside the `if` statement, that means if the `iOSVariant` is not valid, there is nothing going to happen, and there will be no logs, which will make it hard to find and debug issues. 

I would suggest that we keep the `throw` statement but add it in a `else` statement.

Reply:
you're right, but in this form it would not get thrown in case of caught exception from `try` block. What would make sense for me is move throw of exception back outside of `if` block and adding `return;` to end of try block. You decide.

Reply:
adding a `return` statement will work too. Let's keep it that way.